src/Futhark/Pass/ExplicitAllocations.hs

```diff
@@ -733,7 +733,7 @@ addResCtxInIfBody ifrets (Body _ bnds res) spaces substs = do
     mapM_ addStm bnds
     (val_res', mem_ctx_res, bodyrets, _) <-
       foldM helper ([], [], [], ext_ini) (zip4 ifrets val_res substs spaces)
-    return (ctx_res <> mem_ctx_res <> val_res', bodyrets)
+    return (ctx_res `union` mem_ctx_res <> val_res', bodyrets)
```

Comment:
This seems risky to me, as the values returned from the context must match exactly the order in which they occur in the pattern (and the existentials in whatever type is matched to this body).  I have a hard time seeing that treating this as a set can be correct.